session.go

```diff
@@ -0,0 +1,233 @@
+	sessionStateConnected    sessionState = iota
+	sessionStateDisconnected sessionState = iota
+)
+
+type session struct {
+	b *buffer
+
+	conn  Conn
+	connm sync.Mutex
+
+	st  sessionState
+	stm sync.Mutex
+
+	id  uint32
+	idm sync.Mutex
+
+	err SessionError
+}
+
+func NewSession(conn Conn) Session {
+	return &session{
+		st:    sessionStateNew,
+		stm:   sync.Mutex{},
+		conn:  conn,
+		connm: sync.Mutex{},
+		idm:   sync.Mutex{},
+		b:     newBuffer(50),
+	}
+}
+
+func (s *session) Connect() error {
+	if s.isNew() {
+		return errors.New("can't connect unless the session is new")
+	}
+
+	go s.readErrors()
+	return nil
+}
+
+func (s *session) isNew() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st != sessionStateNew
+}
+
+func (s *session) Disconnected() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st == sessionStateDisconnected
+}
+
+func (s *session) Connnected() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st == sessionStateConnected
+}
+
+func (s *session) Send(n Notification) error {
+	// If disconnected, error out
+	if s.Connnected() {
+		return errors.New("not connected")
+	}
+
+	// Set identifier if not specified
+	n.Identifier = s.determineIdentifier(n.Identifier)
+
+	// Serialize
+	b, err := n.ToBinary()
+	if err != nil {
+		return err
+	}
+
+	// Add to buffer
+	s.b.Add(n)
+
+	// Send synchronously
+	return s.send(b)
+}
+
+func (s *session) send(b []byte) error {
+	s.connm.Lock()
+	defer s.connm.Unlock()
+
+	_, err := s.conn.Write(b)
+	if err == io.EOF {
+		s.Disconnect()
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *session) Disconnect() {
+	// Disconnect
+	s.transitionState(sessionStateDisconnected)
+}
+
+func (s *session) RequeueableNotifications() []Notification {
+	notifs := []Notification{}
+
+	// If still connected, return nothing
+	if s.st != sessionStateDisconnected {
+		return notifs
+	}
+
+	// Walk back to last known good notification and return the slice
+	var e *list.Element
+	for e = s.b.Front(); e != nil; e = e.Next() {
+		if n, ok := e.Value.(Notification); ok && n.Identifier == s.err.Notification.Identifier {
+			break
+		}
+	}
+
+	// Start right after errored ID and get the rest of the list
+	for e = e.Next(); e != nil; e = e.Next() {
+		n, ok := e.Value.(Notification)
+		if !ok {
+			continue
+		}
+
+		notifs = append(notifs, n)
+	}
+
+	return notifs
+}
+
+func (s *session) transitionState(st sessionState) {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	s.st = st
+}
+
+func (s *session) determineIdentifier(n uint32) uint32 {
+	s.idm.Lock()
+	defer s.idm.Unlock()
+
+	// If the id passed in is 0, that means it wasn't
+	// set so get the next ID. Otherwise, set it to that
+	// identifier.
+	if n == 0 {
+		s.id++
+	} else {
+		s.id = n
+	}
+
+	return s.id
+}
+
+func (s *session) readErrors() {
+	p := make([]byte, 6, 6)
+
+	_, err := s.conn.Read(p)
+	// TODO(bw) not sure what to do here. It's unclear what errors
```

Comment:
looking into `(c *Conn) Read(b []byte)` in tls/conn.go:

* There are a bunch of "tls: " errors during a handshake on the first read, but they are just error strings without exported Err constants to compare to.
* There is some mention of EAGAIN or EOF errors.
* There's also an io.ErrNoProgress error.